common/sdk/factory.go

```diff
@@ -72,24 +75,33 @@ func NewClientFactory(hostPort string, tlsConfig *tls.Config, metricsHandler *Me
 		hostPort:       hostPort,
 		tlsConfig:      tlsConfig,
 		metricsHandler: metricsHandler,
-		once:           &sync.Once{},
 	}
 }
 
 func (f *clientFactory) NewClient(namespaceName string, logger log.Logger) (sdkclient.Client, error) {
+	f.lock.Lock()
+	defer f.lock.Unlock()
```

Comment:
can we get away without this lock? 
For example, use once.Do(() => firstClient = init_first_client)

Reply:
Maybe? What should the body of the Once do if creating the client fails? There's no way to "reset" the Once and let it try again next time. The Once in `GetSystemClient` can logger.Fatal in that case but I don't think we can do that here. Or can we?

Reply:
oh, no, i didn't realize the creation could fail. ok forget about this.

Reply:
actually it might be reasonable to do that: we can enforce that the client for the system namespace is always created first, and we create others from that one.

as far as I can tell, sdkclient.Dial only needs to make one rpc to GetSystemInfo to succeed. it doesn't even need the requested namespace to exist yet.

and we already call GetSystemClient on startup. so we can just enforce that. let me give it another try...